c++/src/kj/buffering.h

```diff
@@ -0,0 +1,237 @@
+#pragma once
+
+#include <kj/common.h>
+#include <kj/async.h>
+#include <kj/io.h>
+#include <kj/async-io.h>
+#include <kj/one-of.h>
+#include <kj/debug.h>
+
+KJ_BEGIN_HEADER
+
+namespace kj {
+
+// =======================================================================================
+// Buffered reading helpers
+
+struct FdRangeInputBufferOptions {
+  size_t size;
+  // Size of the buffer in bytes.
+
+  size_t maxFds;
+  // Max number of FDs allowed in a single underlying read.
+
+  double retainRatio = 0.8;
+  // When a `RetainableRange` occupies more than `retainRatio` of the buffer,
+  // `retain()` will transfer ownership of the buffer and allocate a new one instead
+  // of performing a copy.
+};
+
+struct EndOfFile{};
+class RetainableRange;
+
+namespace _ {
+
+class FdRangeInputBufferBase {
+protected:
+  FdRangeInputBufferBase(FdRangeInputBufferOptions options);
+
+  FdRangeInputBufferOptions options;
+  Array<byte> allBytes;
+  Array<AutoCloseFd> fds;
+
+  struct Empty {
+    bool eof = false;
+  };
+  struct FdRange {
+    ArrayPtr<byte> bytes;
+    Array<AutoCloseFd> fds;
+  };
+  struct TwoFdRange {
+    ArrayPtr<byte> bytes;
+    Array<AutoCloseFd> oneFds;
+    Array<AutoCloseFd> twoFds;
+    size_t lengthOfFirstRange;
+  };
+  OneOf<Empty, FdRange, TwoFdRange> state = Empty{};
+
+  ArrayPtr<byte> getByteRange();
+  void setByteRange(ArrayPtr<byte> bytes);
+  bool overflowed();
+
+  struct ReadParams {
+    void* start;
+    size_t max;
+  };
+  ReadParams prepareForRead(size_t min);
+  void handleReadResult(size_t byteCount, Array<AutoCloseFd> fds);
+
+  Own<RetainableRange> split(size_t splitAt);
+
+  friend class ::kj::RetainableRange;
+};
+
+} // namespace _
+
+template<typename Stream>
```

Comment:
I don't think this should be a template. It looks like the purpose is just to support AsyncInputStream + AsyncCapabilityStream. I would recommend using a `OneOf` or `Maybe<AsyncCapabilityStream&>` instead. Using a template for this is overkill: it's quite expensive in terms of code bloat (which often translates into poor performance due to instruction cache inefficiency) and templated code is generally harder to work with (e.g. since the IDE doesn't know what the types are, compiler errors are confusing, etc.).